proposals/010-authorizer.md

```diff
@@ -0,0 +1,212 @@
+
+### File based Authorizer implementation
+
+The initial scope will include an Authorizer implementation that is backed by authorization rules expressed in a separate file.  This file will associate principals with ACL rules capable of expressing an allow-list of resources.
+The initial version will be restricted to expressing allow lists of topics, but future version will extend this to allow for rules to be express about other resource types.
+
+### APIs
+
+#### Authorizer Filter
+
+Filter Configuration:
+
+```yaml
+type: AuthorizerFilter
+config:
+  authorizer: FileBasedAllowListAuthorizer
+  authorizerConfig:
+    rulesFile: /path/to/allow-list.yaml
+```
+
+Java APIs:
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.Authorizer, except is asynchronous in nature.
+interface Authorizer {
+    CompletionStage<List<AuthenticationResult>> authorize(AuthorizableRequestContext context, List<Action> actions);
+}
+```
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.AuthorizableRequestContext
+interface AuthorizableRequestContext {
+    String principal();
+    // scope for methods such as requestType(), requestVersion()  etc to be added in future.
+}
+```
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.Action
+record Action(
+    AclOperation aclOperation,
+    ResourcePattern resourcePattern) {
+}
+```
+
+```java
+// The following types are inspire by the Kafka classes of the same name and have the same role.  However, interfaces are used
+// rather than enums to allow for extensibility (using pattern suggested by https://www.baeldung.com/java-extending-enums)
+interface AclOperation {
+    String operationName();
+}
+
+enum CoreAclOperation implements AclOperation {
+  CREATE("Create"),
+  DELETE("Delete"),
+  READ("Read")/* ,... */
+}
+
+interface ResourceType {
+    String resourceType();
+}
+
+
+enum CoreResourceType implements ResourceType {
+  TOPIC("Topic"),
+  CLUSTER("Cluster")
+}
+
+interface PatternType {
+   boolean matches(String resourceName);
+}
+
+
+enum CorePatternType {
+   LITERAL() {
+      @Override
+      boolean matches(String pattern, String resourceName) {
+         return pattern.equals(resourceName);
+      }
+   },
+   MATCH() { /* ... */ },
+   PREFIXED { /* ... */ }
+}
+
+record ResourceNamePattern(PatternType patternType, String pattern) {
+    boolean matches(String resourceName) {
+       return patternType.matches(pattern, resourceName);  
+    }
+}
+```
+
+#### Rules File
+
+The rules file expresses a mapping between principals (user type only with exact match) and an allow-list of resources.
+
+For the initial scope, only resource rules of type TOPIC are supported.  In order to allow for future extension, the user’s configuration set the version property to 1.  This will allow future versions of the filter to introduce support for other resource types without changing the meaning of existing configurations.
```

Comment:
There's something important here which you're not calling out clearly.

1. The rules are deny by default (this is implied by your saying it's always an allow list)
2. We're only supporting Topic resource initially
3. Thus without the version are could not compatibly add a new resource type, because existing rules files would be interpretted as denying access to the newly-supported resource types.

I think it's worth spelling it out more clearly.

Reply:
Done.